### PR TITLE
Catchup at part 1

### PR DESCRIPTION
**Pull request by vogel**
First part of catchup-at implementation - small refactorings

### CODE REVIEW

File: src/catchup/ApplyBucketsWork.cpp [hunk 1]
```diff
@@ -21,25 +21,24 @@ namespace stellar
...
     std::map<std::string, std::shared_ptr<Bucket>>& buckets,
-    HistoryArchiveState& applyState,
-    LedgerHeaderHistoryEntry const& firstVerified)
+    HistoryArchiveState applyState, LedgerHeaderHistoryEntry applyAt)
     : Work(app, parent, std::string("apply-buckets"))
     , mBuckets(buckets)
-    , mApplyState(applyState)
-    , mFirstVerified(firstVerified)
+    , mApplyState(std::move(applyState))
+    , mApplyAt{std::move(applyAt)}
```
Comment (MonsieurNicolas) [inline src/catchup/ApplyBucketsWork.cpp]: why ditching the mFirstVerified name?  Maybe it's not the best name but mApplyAt seems to convey less information. Also you missed a few references to the old name (firstVerified)

**Comment by MonsieurNicolas** [comment]
I started to look at those changes but I think the approach is sub-optimal here and we end up with code churn that doesn't actually solve the root problem that you were going after.

From what I can tell, this PR is really about making sure that we don't have Work objects stepping on each other's state as they perform their work but it doesn't actually succeed in doing so.

In your PR comments you're saying that
> "Also with non-obvious lifetime of Work classes it is dangerous to pass references to them, as Work objects may outlife the references."

In this case it's actually quite safe to hold a reference to a parent Work item as there is an assumption all over the code that we never destroy a work item that has child items. In fact, it makes the code quite easy to follow as it's obvious who owns what (in most cases, see below).

In fact, if copying objects actually helped in anything it would be in hiding fundamental logic bugs in the code as we should not, for example, apply buckets if we're still not sure that we should snap to that state.

I would rather see this PR be revisited to actually make sure that the code's assumptions are correct, namely:
 * some work items assume that there is only one child running at a time (sequence), but right now nothing in the code stops somebody from kicking off work items that will run in parallel to each other (ie: we should enforce that addChild cannot be called if there is already a job in there), as part of that we may have to stop tracking completed/inactive work items in the queue which should help with readability.
 * we should have code in WorkParent that ensures that when a child Work object is being retired it puts it in a state that will cause an assert (will allow to catch issues like in CatchupRecentWork::onReset that is error prone as it clears children but may forget fields pointing to Work).
 * it's unclear to me if we have cases where we respawn duplicate jobs when we should not
    * there is some logic to avoid this (duplicate job detection in WorkParent) but
        * it's unclear to me if it's possible for the same child Work to be spawned from two different parents (we share some code between things)
        * I am not sure we do the right thing in places where we use generated "unique" names for the Work (bypassing the check). For example CatchupTransactionsWork is created with a dynamic name but I don't think we should be running two of these in parallel *ever* (it creates ApplyLedgerChainWork under the cover) - note that at least we would crash if the check I am proposing of detecting uniqueness globally was implemented, right now it would go unnoticed as they would be spawned from different parents.
  * in some cases, I think it may make sense to move ownership of certain data structures to the object doing the actual work, but so far I didn't see any as I like to think of the parent work as the "main method" (that is supposed to hold "temp" variables that cross work item boundaries).

There might be other changes that should go in there, this is what I could think of as a first pass. Let me know if you have other concerns.

**Comment by vogel** [comment]
Thank you for that comment.

> In this case it's actually quite safe to hold a reference to a parent Work item as there is an assumption all over the code that we never destroy a work item that has child items. In fact, it makes the code quite easy to follow as it's obvious who owns what (in most cases, see below).

I think the least we should do is to change that from assumption to something that is directly visible in the code. What about storing work items as `std::unique_ptr` instead of `std::shared_ptr`?

> In fact, if copying objects actually helped in anything it would be in hiding fundamental logic bugs in the code as we should not, for example, apply buckets if we're still not sure that we should snap to that state.

Currently there is also nothing that prevents such bugs from appearing. I propose using a VerifiedLedgerHeaderHistoryEntry for passing history entries that are verified. This could be done by simple wrapper or by a tagged class (https://www.fluentcpp.com/2016/12/08/strong-types-for-strong-interfaces/ ?). This would make it obvious that ApplyBucketsWork requires verified ledger.

I was also thinking about doing something like SequentialWorkParent, but without any deeper thought put into it.

I'll remove last 4 commits from this PR for now.

**Comment by MonsieurNicolas** [comment]
> I think the least we should do is to change that from assumption to something that is directly visible in the code. What about storing work items as std::unique_ptr instead of std::shared_ptr?
I am not sure I am following this comment as the changes were related to the switch from references (`T &mSomeObject`) to copies of objects (`T mSomeObject`).

> Currently there is also nothing that prevents such bugs from appearing. I propose using a VerifiedLedgerHeaderHistoryEntry for passing history entries that are verified. This could be done by simple wrapper or by a tagged class (https://www.fluentcpp.com/2016/12/08/strong-types-for-strong-interfaces/ ?). This would make it obvious that ApplyBucketsWork requires verified ledger.
I would rather make changes to the contracts of `Work` related classes as per my earlier comments so that all Work related classes benefit from the changes. What this would solve doesn't seem to be that much of a problem (ie, we can detect at runtime).

**Pull request closed without merging**